### ISSUE TITLE
Add a parameterless ISpanBuilder.StartActive() overload with the behavior of finishSpanOnDispose:true

### ISSUE THREAD

**Issue by ndrwrbgs** [opentracing/opentracing-csharp#73]
`IScope StartActive(bool finishSpanOnClose);` should not the default be to close the spans on start? I'd presume this behavior should be opt-out rather than opt-in?

**Comment by cwe1ss** [comment]
What do you mean by "close the spans **on start**"? 

As this is a required parameter this is neither opt-in nor opt-out IMO. The user always has to deliberately decide which behavior he/she wants.

**Comment by ndrwrbgs** [comment]
Apologies, miswording, I meant on Dispose.

> The user always has to deliberately decide which behavior he/she wants.

This is what I am disagreeing with.

I would say that not closing the spans is an advanced topic for users -- and in that case the Dispose is a NOOP -- (side note that this means the user can still choose to do this just by not calling `IDisposable.Dispose`).

I do see that the latest version of OT-java removed this, but this seems like a step in the wrong direction. I foresee many folks asking why this input is required and why they would ever not want to close a span, and then we would have to explain more advanced tracing scenarios.

That said, I realize that it's easy to add an extension method locally that removes the verbosity of the calls, so this can just be considered a feature-consider.

**Comment by ndrwrbgs** [comment]
Indeed, upon further reflection I recall that I had taken the approach of easy-to-implement for this project, given the availability of extension methods to fulfill easy-to-utilize (e.g. the Semantics stuff).

That means I'm happy with as-is for a tracer point of view, I'd rather implement one method than have a default parameter or implement two, but I would question how we will ship such utilities to end users. They obviously shouldn't be in the main tracer-implementor-focused package, but I would gander that things like the Semantics or extension methods should be easily available to end users and from an official source.

Without getting into what should and should not be in this package, @cwe1ss do you have any ideas on how such a "consumer focused" package will work in the OT work - both in C# specifically and in the overall OT spec area?

**Comment by cwe1ss** [comment]
Understood. I agree that using `finishSpanOnDispose: true` will be used in the majority of cases and that it could be the default.

We have to decide how we handle such cases. Does it make sense to propose and discuss this in opentracing-java to keep feature parity or is this something that's unique to C# (due to async/await) and therefore an acceptable deviation?

I wouldn't say that this library is tracer-implementor-focused. It's somewhere in the middle and should try to satisfy both worlds and be easy-to-use for instrumentors as well. While it would certainly be possible to implement this as an extension method in an [opentracing-contrib](https://github.com/opentracing-contrib) we have to think about where it fits better. If it's used by everyone anyway, it's probably better to have it directly in the Core API. 

If we end up with a contrib project that contains many deviations and everyone is using them, we might risk ending up with a de-facto-API that might collide with future changes in Java's API.

**Comment by ndrwrbgs** [comment]
> I wouldn't say that this library is tracer-implementor-focused. It's somewhere in the middle and should try to satisfy both worlds and be easy-to-use for instrumentors as well. While it would certainly be possible to implement this as an extension method in an opentracing-contrib we have to think about where it fits better. If it's used by everyone anyway, it's probably better to have it directly in the Core API.

I agree with this whole-heartedly. I rather intended to express that if anything, I think omitting the flag should be done in an extension method rather than in the interface itself, in order to reduce the implementation surface area for tracers.

Tangential but related - from the above what are your feelings about where `Semantics` code should live?

**Comment by cwe1ss** [comment]
@bhs @tedsuo @yurishkuro - I'd appreciate your feedback on this:

Quick summary of this issue: Since C# has `async/await` it's very rare that people have to use `finishSpanOnDispose: false` when calling `StartActive`. In most cases we just use a simple `using` construct and C# takes care of the rest:
```csharp
using (IScope scope = tracer.BuildSpan("foo").StartActive(finishSpanOnDispose: true))
{
    // The C# runtime invokes the SendAsync operation and waits until the response is available.
    // In the meantime all state is stored in an internal state machine 
    // and the thread is available for other processing.
    HttpResponseMessage response = await httpClient.SendAsync(request);

    // As soon as the response is available, the runtime restores the state
    // and execution continues here. 
    Tags.HttpStatus.Set(scope.Span, response.StatusCode)
}
```
Since this is the main scenario and to reduce confusion for users, we would like to add a parameterless overload `StartActive()` that calls `StartActive(finishSpanOnDispose: true)`.

Do you think we can just do this as this is special to C# or should this only be implemented if Java gets the overload as well?

**Comment by yurishkuro** [comment]
Side note - C# is funny: "The async and await keywords were introduced in **Visual Studio 2012**". So IDE release == new language features 😈

**Comment by yurishkuro** [comment]
on a serious note (I couldn't find it in the doc after a quick search) - how does `await` knows that the Scope constitutes the the state? Do they capture all locals / stack?

**Comment by ndrwrbgs** [comment]
@yurishkuro Yes, kind of. For all intents and purposes you can kind of simplify async-await into just synchronous code that maybe pauses sometimes, but as far as the coding itself goes you don't have to worry about that. It's actually a bit confusing when you start using things like `AsyncLocal` and realize that even though your code will run in parallel, you can write it as if it is fully single-threaded. #weird....

**Comment by cwe1ss** [comment]
Yes, the compiler converts an `async` method into a state machine that captures all locals. If you're interested, [here's a good blog post](https://www.dynatrace.com/news/blog/behind-net-4-5-async-scene-performance-impact-asynchronous-programming-c/) that explains it and also shows how the compiler transforms the code.

`AsyncLocal` is similar to Java's (and C#'s) `ThreadLocal` and can be used to store data in the execution context. However, an asynchronous call could be continued on a different thread so `AsyncLocal` keeps the data across threads. [This page](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=netframework-4.7.1#Examples) shows the difference.

We use `AsyncLocal` in our `AsyncLocalScopeManager` to store the `Active` scope.

**Comment by yurishkuro** [comment]
Thanks!

> we would like to add a parameterless overload `StartActive()` that calls `StartActive(finishSpanOnDispose: true)`.

+1

**Comment by ndrwrbgs** [comment]
@yurishkuro 
> Side note - C# is funny: "The async and await keywords were introduced in Visual Studio 2012". So IDE release == new language features

They've tried more recently to decouple the two which is nice, but for the longest time this was indeed true :)

**Comment by bhs** [comment]
@cwe1ss 

> Do you think we can just do this as this is special to C# or should this only be implemented if Java gets the overload as well?

I feel strongly that we should do what makes sense for C#; it's not the same language as Java, so there's no need to have precise 1:1 parity on every method signature. If the `using` idiom is actually the dominant calling pattern, what you're suggesting makes perfect sense to me.

(Sorry for my delay in responding!)

**Comment by cwe1ss** [comment]
thanks for your responses. @ndrwrbgs do you want to create a PR?

**Comment by carlosalberto** [comment]
One interesting scenario would be Java *eventually* using `finishSpanOnClose=false` as the default - in which case, people using different platforms/languages would have to remember what's the default value on each one.

At the same time, I guess the eventual inclusion of `ScopeManager` in the actual specification could mention that default values for `finishSpanOnClose` can vary among supported platforms?

**Comment by carlosalberto** [comment]
Also, I'm wondering if `ScopeManager.Activate()` should have a overload with `finishSpanOnDispose=true` (as we had it in Java in the past as well).

**Comment by cwe1ss** [comment]
We've briefly discussed this in the cross-language working group. We want to make sure we don't end up with different languages having different defaults so we want to properly discuss this in the specification repository.

Please have a look at https://github.com/opentracing/specification/pull/108

I'll put this on hold until that specification-issue is resolved.

**Comment by carlosalberto** [comment]
FYI, we already have integrated this as part of the second RC for `opentracing-python` 2.0. https://github.com/opentracing/specification/pull/108 has been merged but is still on the works, but the part of the default default value for `finishSpanOnClose` has been covered there already:

> A default value for this parameter may be provided, depending on the suitability for the language and its use cases.

For your consideration on merging this PR ;)

**Comment by cwe1ss** [comment]
@carlosalberto thanks for the update! This sounds good, so we'll add it as well! It really makes sense for C#!!
